core/preconditioner/jacobi.cpp

```diff
@@ -231,11 +231,23 @@ void Jacobi<ValueType, IndexType>::generate(const LinOp *system_matrix,
     using csr_type = matrix::Csr<ValueType, IndexType>;
     const auto exec = this->get_executor();
     if (parameters_.max_block_size == 1) {
-        auto diag = as<DiagonalExtractable<ValueType>>(system_matrix)
-                        ->extract_diagonal();
-        auto temp = gko::Array<ValueType>::view(
-            exec, system_matrix->get_size()[0], diag->get_values());
-        this->blocks_ = temp;
+        auto diag = share(as<DiagonalLinOpExtractable>(system_matrix)
+                              ->extract_diagonal_linop());
+        using next_type = next_precision<ValueType>;
+        if (auto diag_dense =
+                std::dynamic_pointer_cast<matrix::Diagonal<ValueType>>(diag)) {
+            auto temp = gko::Array<ValueType>::view(
+                exec, system_matrix->get_size()[0], diag_dense->get_values());
```

Comment:
The source of the segfault we just saw:
```suggestion
                diag_dense->get_executor(), system_matrix->get_size()[0], diag_dense->get_values());
```

Reply:
good catch!

Reply:
Could you add a test for this behavior (generating with block size 1 on CUDA from a Reference/OMP matrix)?

Reply:
Good catch. I think we should think about having a better way to test these cross-executor combinations ? I think we hardly test any cross-executor functionality now.

Reply:
for easy test, we can use `get_test_executor()` in random or in order to avoid specify the executor in each place

Reply:
Yeah, I think you can do a specific test for this and we can think discuss a more generic approach later.